src/Middleware/HttpLogging/src/FileLoggerProcessor.cs

```diff
@@ -0,0 +1,269 @@
+            _path = loggerOptions.LogDirectory;
+            if (string.IsNullOrEmpty(_path))
+            {
+                _path = Path.Join(environment.ContentRootPath, "logs", DateTimeOffset.Now.ToUnixTimeSeconds().ToString(CultureInfo.InvariantCulture));
+            }
+            _fileName = loggerOptions.FileName;
+            _maxFileSize = loggerOptions.FileSizeLimit;
+            _maxRetainedFiles = loggerOptions.RetainedFileCountLimit;
+            _flushInterval = loggerOptions.FlushInterval;
+            _options.OnChange(options =>
+            {
+                // Clear the cached settings.
+                loggerOptions = options;
+                if (!string.IsNullOrEmpty(loggerOptions.LogDirectory))
+                {
+                    _path = loggerOptions.LogDirectory;
+                }
+                _fileName = loggerOptions.FileName;
+                _maxFileSize = loggerOptions.FileSizeLimit;
+                _maxRetainedFiles = loggerOptions.RetainedFileCountLimit;
+                _flushInterval = loggerOptions.FlushInterval;
+            });
+
+            _logger = factory.CreateLogger(typeof(FileLoggerProcessor));
+
+            // Start message queue processor
+            _cancellationTokenSource = new CancellationTokenSource();
+            _outputTask = Task.Run(ProcessLogQueue);
+        }
+
+        public void EnqueueMessage(string message)
+        {
+            if (!_messageQueue.IsAddingCompleted)
+            {
+                try
+                {
+                    _messageQueue.Add(message);
+                    return;
+                }
+                catch (InvalidOperationException) { }
+            }
+        }
+
+        private async Task ProcessLogQueue()
+        {
+            while (!_cancellationTokenSource.IsCancellationRequested)
+            {
+                while (_messageQueue.TryTake(out var message))
+                {
+                    _currentBatch.Add(message);
+                }
+                if (_currentBatch.Count > 0)
+                {
+                    try
+                    {
+                        await WriteMessagesAsync(_currentBatch, _cancellationTokenSource);
+                    }
+                    catch (Exception ex)
+                    {
+                        Log.WriteMessagesFailed(_logger, ex);
+                    }
+
+                    _currentBatch.Clear();
+                }
+                else
+                {
+                    try
+                    {
+                        await Task.Delay(_flushInterval, _cancellationTokenSource.Token);
+                    }
+                    catch
+                    {
+                        // Exit if task was canceled
+                        return;
+                    }
+                }
+            }
+        }
+
+        private async Task WriteMessagesAsync(List<string> messages, CancellationTokenSource cancellationTokenSource)
```

Comment:
Nit: All the CancellationTokenSource parameters should be CancellationTokens instead. The CancellationTokenSource should only be used for cancelling in Dispose.